File: app/src/main/java/com/example/cinemates/view/ui/profile/ProfileFragment.java
package com.example.cinemates.view.ui.profile;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.cinemates.adapter.SectionRecyclerViewAdapter;
import com.example.cinemates.databinding.FragmentProfileBinding;
import com.example.cinemates.model.data.Movie;
import com.example.cinemates.model.data.Person;
import com.example.cinemates.model.data.Section;
import com.example.cinemates.util.ViewSize;
import com.example.cinemates.view.viewmodel.DbViewModel;
import com.google.android.material.transition.MaterialFadeThrough;

import java.util.ArrayList;
import java.util.List;

public class ProfileFragment extends Fragment {

    private FragmentProfileBinding mBinding;
    private Section<Movie> mMovieSection;
    private Section<Person> mPersonSection;
    private DbViewModel mDbViewModel;
    private SectionRecyclerViewAdapter mAdapter;
    private List<Section<?>> mSectionList;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAdapter = new SectionRecyclerViewAdapter(this, getContext());
        mMovieSection = new Section<>("Favorites", "Movies", Movie.class, null, ViewSize.SMALL);
        mPersonSection = new Section<>("Favorites", "Actors", Person.class, null, ViewSize.SMALL);
        mSectionList = new ArrayList<>();
        mSectionList.add(mMovieSection);
        mSectionList.add(mPersonSection);
        mDbViewModel = new ViewModelProvider(getActivity()).get(DbViewModel.class);
        setupMotionAnimations();
    }

    private void setupMotionAnimations() {
        setEnterTransition(new MaterialFadeThrough());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mBinding = FragmentProfileBinding.inflate(inflater, container, false);
        mBinding.recyclerView.setAdapter(mAdapter);
        mAdapter.addItems(mSectionList);
        mBinding.statHours.statTitle.setText("Hours");
        mBinding.statWatchedCounter.statTitle.setText("Movies Seen");
        mBinding.statToSeeCounter.statTitle.setText("Movies To See");
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mMovieSection.setMutableLiveData(mDbViewModel.getFavorite_movies());
        mPersonSection.setMutableLiveData(mDbViewModel.getFavorite_persons());
        mDbViewModel.getAllFavoritesMovies();
        mDbViewModel.getAllFavoritesPersons();
        mDbViewModel.getAllWithStatus(Movie.PersonalStatus.SEEN);
        mBinding.statHours.statContent.setText(formatTime(mDbViewModel.sumRuntimeAllWatchedMovies()));
        mBinding.statWatchedCounter.statContent.setText(String.valueOf(mDbViewModel.getMovieCountByStatus(Movie.PersonalStatus.SEEN)));
        mBinding.statToSeeCounter.statContent.setText(String.valueOf(mDbViewModel.getMovieCountByStatus(Movie.PersonalStatus.TO_SEE)));

    }

    private String formatTime(long runtime) {
        long hours = (long) (runtime / 60);
//        int minutes = (int) (runtime % 60);
        return String.valueOf(hours);

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mBinding = null;
    }
}

File: app/src/main/java/com/example/cinemates/view/ui/profile/ProfileFragment.kt
package com.example.cinemates.view.ui.profile

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import com.example.cinemates.R
import com.example.cinemates.adapter.SectionRecyclerViewAdapter
import com.example.cinemates.databinding.FragmentProfileBinding
import com.example.cinemates.databinding.LayoutSectionStatsBinding
import com.example.cinemates.model.data.Movie
import com.example.cinemates.model.data.Person
import com.example.cinemates.model.data.PersonalStatus
import com.example.cinemates.model.data.Section
import com.example.cinemates.util.ViewSize
import com.example.cinemates.view.viewmodel.DbMovieViewModel
import com.example.cinemates.view.viewmodel.DbPersonViewModel
import com.example.cinemates.util.getLong
import com.google.android.material.transition.MaterialFade
import com.google.android.material.transition.MaterialFadeThrough
import com.google.android.material.transition.MaterialSharedAxis

/**
 * @author Antonio Di Nuzzo
 * Created 28/08/2022 at 10:13
 */
class ProfileFragment : Fragment() {
    private var _binding: FragmentProfileBinding? = null
    private val binding: FragmentProfileBinding
        get() = _binding!!
    private val dbMovieViewModel: DbMovieViewModel by activityViewModels()
    private val dbPersonViewModel: DbPersonViewModel by activityViewModels()

    private lateinit var adapter: SectionRecyclerViewAdapter

    //Sections
    private val movieSection: Section<Movie> =
        Section("Favorites", "Movies", Movie::class.java, null, ViewSize.SMALL)
    private val personSection: Section<Person> =
        Section("Favorites", "Actors", Person::class.java, null, ViewSize.SMALL)
    private val sectionList: List<Section<*>> =
        listOf(movieSection, personSection)


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter = SectionRecyclerViewAdapter(this)
        enterTransition = MaterialFadeThrough()
        reenterTransition = MaterialSharedAxis(MaterialSharedAxis.X, false).apply {
            interpolator = FastOutSlowInInterpolator()
            duration = resources.getLong(R.integer.material_motion_duration_long_2)
        }
        exitTransition = MaterialFade()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        // Inflate the layout for this fragment
        _binding = FragmentProfileBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.apply {
            recyclerView.adapter = adapter
            adapter.addItems(sectionList)

            setCounter(statHours, "Total Hours", dbMovieViewModel.getTotalHoursOf(PersonalStatus.EMPTY))
            setCounter(statWatchedCounter, "Movies Seen", dbMovieViewModel.getSizeOf(PersonalStatus.SEEN))
            setCounter(statToSeeCounter, "Movies To See", dbMovieViewModel.getSizeOf(PersonalStatus.TO_SEE))
        }
        movieSection.liveData = dbMovieViewModel.favorites
        personSection.liveData = dbPersonViewModel.persons

    }

    private fun setCounter(stats: LayoutSectionStatsBinding, title:String, value:Int ){
        stats.statTitle.text = title
        stats.statContent.text = value.toString()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }


}